core/field.js

```diff
@@ -619,22 +599,72 @@ Blockly.Field.prototype.updateWidth = function() {
 
 /**
  * Updates the size of the field based on the text.
+ * @param {number=} opt_margin margin to use when positioning the text element.
  * @protected
  */
-Blockly.Field.prototype.updateSize_ = function() {
-  var textWidth = Blockly.utils.dom.getFastTextWidth(
-      /** @type {!SVGTextElement} */ (this.textElement_),
-      this.constants_.FIELD_TEXT_FONTSIZE,
-      this.constants_.FIELD_TEXT_FONTWEIGHT,
-      this.constants_.FIELD_TEXT_FONTFAMILY);
-  var totalWidth = textWidth;
+Blockly.Field.prototype.updateSize_ = function(opt_margin) {
+  var constants = this.constants_;
+  var xOffset = opt_margin != undefined ? opt_margin :
+      (this.borderRect_ ? this.constants_.FIELD_BORDER_RECT_X_PADDING : 0);
+  var totalWidth = xOffset * 2;
+  var totalHeight = constants.FIELD_TEXT_HEIGHT;
+
+  var contentWidth = 0;
+  if (this.textElement_) {
+    contentWidth = Blockly.utils.dom.getFastTextWidth(this.textElement_,
+        constants.FIELD_TEXT_FONTSIZE + 'pt',
+        constants.FIELD_TEXT_FONTWEIGHT,
+        constants.FIELD_TEXT_FONTFAMILY);
+    totalWidth += contentWidth;
+  }
   if (this.borderRect_) {
-    totalWidth += this.constants_.FIELD_BORDER_RECT_X_PADDING * 2;
-    this.borderRect_.setAttribute('width', totalWidth);
+    totalHeight = Math.max(totalHeight, constants.FIELD_BORDER_RECT_HEIGHT);
   }
+
+  this.size_.height = totalHeight;
   this.size_.width = totalWidth;
+
+  this.positionTextElement_(xOffset, contentWidth);
+  this.positionBorderRect_();
 };
 
+/**
+ * Position a field's text element after a size change.
+ * @param {number} xOffset x offset to use when positioning the text element.
+ * @param {number} contentWidth The content width.
+ * @protected
+ */
+Blockly.Field.prototype.positionTextElement_ = function(xOffset, contentWidth) {
+  if (!this.textElement_) {
+    return;
+  }
+  var constants = this.constants_;
+  var halfHeight = this.size_.height / 2;
+
+  this.textElement_.setAttribute('x', this.sourceBlock_.RTL ?
```

Comment:
Since you're doing updateSize on every render, it seems worth checking whether anything has changed before calling setAttribute (probably in updateSize, not here).

Reply:
Agreed, we would be getting into the territory of keeping a cache of these values on the field itself (which I'm not opposed to).

Reply:
I'm going to take care of this in a separate change.

Reply:
https://github.com/google/blockly/issues/3701